### PR TITLE
fix(state): Fix the deduplication of note commitment trees

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation & Solution

- Refactor the deduplication.
- Include a test that checks that there are no duplicates.

This PR merges #7365.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs [hunk 1]
```diff
@@ -272,180 +275,149 @@ impl DbFormatChange {
...
-                let mut delete_from = sapling_height.next();
-                while let Ok(delete_to) = unique_sapling_tree_height_rx.recv() {
-                    let num_entries = delete_to - delete_from;
-                    if num_entries > 0 {
+                // We delete duplicate trees in batches. A batch is a (possibly empty) series of
+                // identical trees excluding the first tree. We exclude the first tree so that we
+                // don't delete it. We get a batch if we encounter a tree that differs from the
+                // previous one, or if we're at the end of the chain. We get an empty batch if there
+                // are two consecutive differing trees.
+                if last_tree != tree || height == initial_tip_height {
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: I think the delete range needs to be different for these two different cases:
1. if the trees aren't equal, we want to keep the tree at `height`
2. if the trees are equal but we've reached the initial tip height, we want to delete the tree at `height`

The code below implements the first case.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs [hunk 2]
```diff
@@ -272,180 +275,149 @@ impl DbFormatChange {
...
-                let mut delete_from = orchard_height.next();
-                while let Ok(delete_to) = unique_orchard_tree_height_rx.recv() {
-                    let num_entries = delete_to - delete_from;
-                    if num_entries > 0 {
+                // We delete duplicate trees in batches. A batch is a (possibly empty) series of
+                // identical trees excluding the first tree. We exclude the first tree so that we
+                // don't delete it. We get a batch if we encounter a tree that differs from the
+                // previous one, or if we're at the end of the chain. We get an empty batch if there
+                // are two consecutive differing trees.
+                if last_tree != tree || height == initial_tip_height {
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: Same here

**Review by teor2345 — COMMENTED** [review]
Arya told me there were some performance issues in the initial version of this PR. Have they been fixed? How fast is this new upgrade code on a v25.0.0 state?

(I can test this if you don't have an old state handy.)

If there are still performance issues, we might want to see if it helps to optimise NoteCommitmentTree::eq() to avoid re-calculating the root if it hasn't been cached in memory or on disk.

Because I think previous state versions didn't cache/serialize the root for empty trees? Or was that just the genesis tree?

**Comment by teor2345** [comment]
The checkpoint test failure in CI:
https://github.com/ZcashFoundation/zebra/actions/runs/5965091001/job/16184178963?pr=7379
should be fixed by my PR #7385.

I'm not sure about the RPC conflict test. It might be a performance issue with the upgrade, so I'll push a fix PR to make the upgrade faster. If that doesn't work, we might need to increase the wait time in that test.

**Comment by teor2345** [comment]
PR #7386 might fix the RPC conflict test failure:
https://github.com/ZcashFoundation/zebra/actions/runs/5965091001/job/16183944554?pr=7379

**Comment by teor2345** [comment]
With this PR, #7385, and #7386, it takes about 35 seconds on my machine to upgrade a fully synced mainnet state and do the runtime tests.

The sync is launched concurrently, but the network hasn't even fully started by then, so I think that performance is acceptable.

On testnet it takes about 10 seconds.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 2 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: This might be the cause of the panic I get on my testnet state:

```
The application panicked (crashed).
Message:  Zebra could not prune all Orchard trees.
Location: zebra-state/src/service/finalized_state/disk_format/upgrade.rs:409

Metadata:
version: 1.1.0+62.g961d45c
Zcash network: Testnet
running state version: 25.1.1
initial disk state version: 25.1.0
features: default,getblocktemplate_rpcs,metrics_exporter_prometheus,prometheus,release_max_level_info 
branch: state-dedup-fixes
git commit: 961d45c
```

You might want to turn this panic into an error log, because the detailed checks show which trees are still duplicates, but this panic doesn't.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 1 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: When I run using all 3 PRs on an old state that was synced to 1719530, I get a panic at the initial tip height:
```
2023-08-25T03:06:36.284392Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate sapling trees after running de-duplicate tree upgrade height=Height(1719530) prev_height=Hei
ght(1719517) tree_root=Root("d6efc94648b492e0a8ee310f4032aa618b565cc124e7cc858137f2f36fdb4e70")
The application panicked (crashed).
Message:  found duplicate sapling or orchard trees after running de-duplicate tree upgrade
Location: zebra-state/src/service/finalized_state/disk_format/upgrade.rs:486

Metadata:
version: 1.1.0+62.g961d45c
Zcash network: Mainnet
running state version: 25.1.1
initial disk state version: 25.0.0
features: default,getblocktemplate_rpcs,metrics_exporter_prometheus,prometheus,release_max_level_info 
branch: state-dedup-fixes
git commit: 961d45c
```

**Comment by teor2345** [comment]
When I run using all 3 PRs on an old mainnet state that was synced to 1719530, the deduplication takes about 7 seconds.

So I think we're fine for performance here.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 1 above)
Comment (upbqdn) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: This edge case is addressed in [`4b579a6` (#7379)](https://github.com/ZcashFoundation/zebra/pull/7379/commits/4b579a60786751719110aee6fcfdde9262b02aef).

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 2 above)
Comment (upbqdn) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: This edge case is addressed in [`4b579a6` (#7379)](https://github.com/ZcashFoundation/zebra/pull/7379/commits/4b579a60786751719110aee6fcfdde9262b02aef).

**Comment by upbqdn** [comment]
> Arya told me there were some performance issues in the initial version of this PR.

The issue was that I was iterating over the trees using `next_back` for each tree. I never ran the upgrade for more than 12 hours, but it would take way longer than that.

> previous state versions didn't cache/serialize the root for empty trees? Or was that just the genesis tree?

That was just the genesis tree. The rest of the trees were stored with their cached roots.

**Comment by teor2345** [comment]
> > Arya told me there were some performance issues in the initial version of this PR.
> 
> The issue was that I was iterating over the trees using `next_back` for each tree. I never ran the upgrade for more than 12 hours, but it would take way longer than that.
> 
> > previous state versions didn't cache/serialize the root for empty trees? Or was that just the genesis tree?
> 
> That was just the genesis tree. The rest of the trees were stored with their cached roots.

Did we update the genesis tree to have a cached root as part of this upgrade or the previous one?

I don't know how often the genesis root is accessed, but it seems like a quick fix for a significant user-triggerable performance/security issue. (It should probably go in a separate PR so we can get this one merged.)

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs [hunk 3]
```diff
@@ -284,17 +286,30 @@ impl DbFormatChange {
...
+                Arc::new(sapling::tree::NoteCommitmentTree::default()),
+            ))
+            .into_iter();
+
             // Run through all the trees in the finalized chain.
-            for (height, tree) in db.sapling_tree_by_height_range(Height(1)..=initial_tip_height) {
+            for (height, tree) in db
+                .sapling_tree_by_height_range(Height(1)..=initial_tip_height)
+                .chain(dummy_cap)
+            {
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: This looks like a good fix, but if the state hasn't reached the activation height yet, the `initial_tip_height` will remain in the database, because the `dummy_cap` tree is equal to the `initial_tip_height` tree. So neither of these conditions is satisfied:
```rust
if last_tree != tree || height == initial_tip_height { ... }
```

I'll push a suggested refactor to another PR.

Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: Once we fix the versioning in PR #7385, this bug would have been caught by the checkpoint test, because the mandatory checkpoint is before NU5 activation

Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: PR #7391 should fix this, I tested it on my legacy state with the fix in PR #7385.

**Comment by upbqdn** [comment]
> Did we update the genesis tree to have a cached root as part of this upgrade or the previous one?

We didn't.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs [hunk 4]
```diff
@@ -236,8 +241,8 @@ impl DbFormatChange {
         config: Config,
         network: Network,
         initial_tip_height: Option<Height>,
-        upgrade_db: ZebraDb,
-        should_cancel_format_change: Arc<AtomicBool>,
+        db: ZebraDb,
+        cancel_receiver: mpsc::Receiver<CancelFormatChange>,
```
Comment (arya2) [inline zebra-state/src/service/finalized_state/disk_format/upgrade.rs:244]: ~Optional/Nitpick: I think it may be useful to use a cloneable type here.~

Update: it's good as-is.

**Review by arya2 — COMMENTED** [review]
This looks good to me once we merge in #7391.

What's this test failure:

> 2023-08-28T02:53:02.693713Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate sapling trees after running de-duplicate tree upgrade height=Height(25900) prev_height=Height(25899) tree_root=Root("fbc2f4300c01f0b7820d00e3347c8da4ee614674376cbc45359daa54f9b5493e")
https://github.com/ZcashFoundation/zebra/actions/runs/5994564430/job/16257493713#step:8:26063

(This is from the suggestion PR, I think it's not re-running the upgrade because the disk version was updated earlier.)

**Comment by upbqdn** [comment]
> What's this test failure

Yea, it looks like the upgrade didn't run at all since there are duplicates starting at height 1.

**Comment by upbqdn** [comment]
Images in GCP are still labeled with 25.1.0.

**Comment by upbqdn** [comment]
> What's this test failure

That was due to the bug described in PR #7385. Since the cached state hasn't been upgraded, Zebra incorrectly determined it was a new state, and labeled it as 25.1.1, which was the version in the code. Zebra then skipped the pruning since it's for lower versions, but the test that checks for duplicates runs at each start, so it reported duplicates starting from height 1. This should be resolved by merging PR #7385.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 4 above)
Comment (upbqdn) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs:244]: I thought we were using a single consumer for little overhead.

**Review by arya2 — APPROVED** [review]

**Comment by upbqdn** [comment]
@mergifyio update.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs:244]: Just a note in case this ever becomes a priority (but it's not at the moment, because we don't need concurrent upgrades yet):

`std::sync::mpsc::Receiver` takes a shared reference, so it can be called on `Arc<Receiver>`:
https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#method.try_recv

(This is different from `tokio::sync::mpsc::Receiver`, which requires an exclusive reference.)

However, if we were using an `mpsc` channel from multiple threads, only the first thread would get the cancel message. The second thread would keep running until the sender was dropped. (Which doesn't depend on the update thread finishing, so that's ok.)

A correct cloneable implementation would be `matches!(tokio::sync::watch::Receiver::has_changed(), Ok(true) | Err(_))`, with a wrapper that clears the initial change notification:
https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.has_changed

There might also be other ways of doing it.
